### ISSUE TITLE
TypeScript

### ISSUE THREAD

**Issue by peterp** [redwoodjs/redwood#234]
----
This tracking issue is no longer maintained, please reference the project board for the latest updates instead:
https://github.com/redwoodjs/redwood/projects/2
----

We're going to add first-class support for TypeScript and be an even better experience for JavaScript as a result.

**Make the project that's created by `yarn create redwood-app` support TypeScript:**
- [x] Convert the `*.js` files in `create-redwood-app` repo to TypeScript.
- [x] Transpile the new TypeScript files into JavaScript if the target is JavaScript. (JavaScript will be the default.)
- [x] Add a flag to switch the target to TypeScript in the `create-redwood-app` package.
- [x] Check for presence of `tsconfig.json` file in a project

**Make the generators TS/ JS aware:**
👉See sub-Tracking Issue #523 
- [x] Make the generators understand that the target is TS/ JS.
- [x] Rewrite the templates from JS into TS.
- [x] When the target is JS transpile them.

**The api side:**
- [x] Verify that we import `.ts` files via the `importAll.macro` (tests!)
- [x] We auto import `context`, is that a problem in the ts world?
- [x] Verify that the dev server works well with `.ts` files.

**The bridge between api and web:**

- [x] Share type definitions from SDL in api with gql`` in web, fixed via #1321 
- [x] Share the `currentUser` returned from `api/src/lib/user.ts` with `getCurrentUser` in `useAuth`

**The web side:**
- [x] Do we properly handle `.ts/ .tsx` files?
- [x] Do our auto-imported routes work properly with tslint?

**Framework/Packages**
- [x] makeMergedSchema #563 

**Tooling**
- [x] Does linting work?
- [x] Does jest work?

**Comment by Domino987** [comment]
Is there a Branch?

**Comment by mohsen1** [comment]
Are you planning to move the core itself to TS? The router needs to have types to work well in TS projects.

**Comment by thedavidprice** [comment]
Currently in process by Peter:
- https://github.com/redwoodjs/redwood/pull/264
- https://github.com/redwoodjs/create-redwood-app/pull/37

> Are you planning to move the core itself to TS?
Yes indeed. 

I believe @jeremyscatigna might have a start on this already. True, Jeremy?

**Comment by johnslemmer** [comment]
In this adding of "first class typescript support", is there a line item for generating types for gql queries?  Something like apollo-codgen?  This is currently one of the pain points when integrating with any API (whether REST or graphql) and if that would be included in the Redwood framework that would be amazing and make for an awesome and very type-safe coding experience.

**Comment by divideby0** [comment]
+1 on generating types for gql queries. The [graphql-code-generator](https://graphql-code-generator.com/docs/plugins/typescript-react-apollo) project might also worth considering as it generates typesafe React Hooks, HOCs, etc for pre-defined query fragments.

**Comment by marceloverdijk** [comment]
Is there a rough timeline for TypeScript support? E.g months, 1/2 year, 1 year?

**Comment by thedavidprice** [comment]
Hi @marceloverdijk it’s a top 3 priority (on the list with Auth and Jest tests, for example), but the progress has admittedly been slow. I know we _want_ it to be “months”. But that’s all the certainty I can commit to regarding a rough timeline.

What I _could_ commit to, personally, is helping organize a collaborative effort around this. I know:
- @mohsen1 is and will continue to contribute (and has more experience with TS than the entire core team combined),
- there was an early [offer to help on our forum](https://community.redwoodjs.com/t/any-wip-on-typescript-support/165),
- and @Domino987 was looking to help (above)…

…but we haven’t really rounded up critical mass. So taking this as a nudge in that direction, and as a next-step I just this created a Forum topic (below). Chime in you’re available + interested:
https://community.redwoodjs.com/t/team-typescript-for-those-who-want-to-help-add-ts-support/369

**Comment by marceloverdijk** [comment]
Thx @thedavidprice for the honest answer. Much appreciated!
I'm mainly a Java developer, taking some side paths in the evening, like JS/TS. But I'm not a level to contribute unfortunately. But I will follow the thread with a lot of interest.

**Comment by peterp** [comment]
Our development configurations are now "TypeScript aware" via #438, this means that if you rename the `.js` files to `.ts` and add `.tsx` files that the development servers will understand how to work with those - and you should not be personally blocked by developing in TypeScript in a Redwood project.

The next steps are to figure out if building a project is working, and to add types to the redwood router.

We'll add support for starting a redwood project in typescript, and making sure the generators generate TypeScript.

**Comment by aldonline** [comment]
There is *a lot* to discuss here. I was having trouble figuring out where to start, but in the service of moving fast, I'll just write part of my thoughts down.

- If we want to make Redwood TypeScript friendly (and of course we do!),
  then you need to be ensure that **a Redwood project works with the TypeScript compiler out of the box**. In other words, Redwood code MUST be valid TypeScript code.
- There is only one TypeScript compiler, and one set of rules. So the goal is clear. In fact, the TypeScript compiler is what most IDEs use to enable language features (Including code completion for JavaScript!). So enabling TypeScript support has many benefits, even if you don't use TypeScript.
- If Redwood has some "magical" tricks going on, then they need to be mapped into TypeScript. If this fit doesn't come naturally, there are three things that can be done: Change the feature, use code generation or modify the behavior of the TypeScript compiler is via tsconfig.json. The latter is obviously the first option. Unfortunately...
- Configuration via tsconfig.json is very restrictive when compared to webpack/babel. This means that there are many things that can be achieved with babel/webpack that _cannot_ be mapped to TypeScript compiler settings
- There are no TypeScript "plugins" like there are for babel/webpack
  - TypeScript does provide Language Service Plugins and Custom Transformers, but they are not relevant, so I'll leave them out of the discussion for now. Without going into details:
    - Custom Transformers - cannot only be enabled programmatically. This means that the only way to use them is to use an alternative TypeScript compiler (a "wrapped" compiler)
    - Language Service Plugins - These can be enabled easily via node_modules and tsconfig, but they can only affect the editing experience

With this in mind, I can see a few challenges when thinking about "porting" Redwood to the TypeScript world. (beyond the obvious- ex: having the code generators output typescript)

# Challenges

## Using path aliases (src/\*) instead of relative imports

This is actually not a real challenge and is easily solved. I'll just add it here in case anyone is wondering.
This can be easily be represented by settings in `tsconfig.json`, which are actually equivalent to the settings in `jsconfig.json`. (in fact, jsonfig.json is a subset of tsconfig.json)

## Directory Name convention

Redwood uses the DirectoryName convention extensively.
This convention is meant to improve Developer Experience by allowing more significant file names (instead of having multiple files named "index.js")
This is enabled via a webpack plugin and, functionally, it extends webpack's module resolution logic by adding an extra alternative:

For the following import:

```js
import MyComponent from "src/MyComponent"
```

Normal Lookup (simplification)

```
src/MyComponent.(t|j)sx?
src/MyComponent/index.(t|j)sx?
node_modules...
```

With DirectoryName convention:

```
src/MyComponent.(t|j)sx?
src/MyComponent/index.(t|j)sx?
src/MyComponent/MyComponent.(t|j)sx? <---- extra step
node_modules...
```

### Can it be mapped to TypeScript?

- Currently there is no way to do this via tsconfig.json
- Code generation can get us half way: we could inject generated declarations for module "src/MyComponent" so the TypeScript compiler doesn't complain. At runtime we don't change the webpack loader logic, since these are two different things. The only issue here is that "go to definition" would take users to the generated code instead of the desired component.
- The only complete solution would be to provide a Custom TypeScript compiler (more info on this at the end of this comment). Internally, this compiler would provide its own "ts.resolveModuleNames" implementation

## Auto Imports on Routes.js

- Page components are automatically imported and made available to Routes.js

### Possible Solution: Code Generation

If we're already generating code, we could "declare global var"iables for each of these pages.
This would make the identifiers available so the compiler would not complain, but it would not
allow "jump to definition" to work correctly, since it would only take us to the generated declaration.
The runtime bundling magic would still be left to webpack.

### Possible Solution: Routes.js is a special case

- Even if you're writing a TypeScript project, Routes.js must be a .js file.
- We can still add some level of IDE features with specialized extensions. Like we're doing in [Decoupled Studio](https://marketplace.visualstudio.com/items?itemName=decoupled.studio)

## importAll.macro

- If babel is still used for transpiling (instead or in addition to the TypeScript compiler), then this should still work as expected. (challenge: the processing pipeline needs to be set in the right order)
- If we wish the output of calling importAll() to be correctly typed at design time, we can do that via code-generation and overloading, with a few border-cases.

## Cells

- Cells are automatically wrapped as a React Component via a custom webpack loader
- This means that there is a mismatch between the static type of a Cell file module and the type
  you end up using when you import it.
- TypeScript has no way of knowing about this change

Solution?

- With code generation, we could generate a wrapper for each cell
- This generated file could be named in such a way that the TypeScript compiler would pick it up

## Providing a Custom TypeScript Compiler

- If we need to do things that cannot be achieved via tsconfig.json + code generation, then Redwood could provide its own "wrapped/custom" TypeScript compiler
- This is a version of the compiler with minor changes to account for the special features

How can this be distributed?
I've seen two approaches:

### Publish a separate NPM module

- It can be published to NPM as "@redwood/typescript", and added as dev dependency to all redwood projects (instead of the vanilla typescript)
- When using VSCode, for example, users would configure the IDE to use this bundled version
- This is somewhat officially supported (you can tell VSCode which `TypeScript SDK folder` to use via config, or we could automatically output/create a `.vscode/settings.json` file)

Challenges:

- It is obviously a very extreme measure to take. If there is no clear justification, we'll get some backlash from the community.
- Also, other tools (outside of redwood and VSCode) might still look up the "typescript" module by its name (expecting it to be in node_modules/typescript or equivalent).

### Monkey Patch

- We can "monkey patch" the official compiler with a post-installation script.
- This is not as tricky as it sounds. The way the TypeScript compiler codebase is designed allows for some things to be "overriden" easily by injecting a few lines of code that run before the rest of the modules are loaded. Specially module resolution logic.
- Without going into details, the way this works is by replacing tsc.js, server.js, etc with a thin loader that injects some code, which then loads the original files.

Challenges:

- I am not sure if all the use cases can be covered this way
- How reliable are post-install hooks? I have seen them being used for "important" stuff by "important" teams, (including the vscode team using it to download vscode.d.ts until two releases ago).

**Comment by koolamusic** [comment]
Maybe I could take a hit at writing type definitions for <Route>, <Link>, and <Private> at the moment

**Comment by thedavidprice** [comment]
@koolamusic that would be fantastic! Looping in @peterp here.

Maybe open a separate Issue (linking back to this one) to get started? And definitely don't hesitate to let us know what questions + clarification + help you need along the way.

🚀

**Comment by tomitrescak** [comment]
@thedavidprice do you have any developer guidelines for adding types? Do you want to store them in one big file, or separate definitions into linked files? Do you want to keep definitions in Redwood repo, or shall it be tracked in the @types repo?

**Comment by thedavidprice** [comment]
Hi @tomitrescak! Thanks for checking in here. @peterp is the one to provide an update as this issue is (in a good way) out of date. And your question is a bit of "it depends" 😆. For the Framework (this repo redwoodjs/redwood) we have a mix of approaches. For the application development template (redwoodjs/create-redwood-app) there's a lot more customization involved and (I believe) includes storing in a big file.

**Comment by koolamusic** [comment]
@tomitrescak I think it's okay to start out with one file for some packages, while others could use a type-per-module approach to separate the definitions. With regards to the developer guidelines, have you seen https://github.com/sindresorhus/typescript-definition-style-guide ? would love to know what your thoughts are

**Comment by peterp** [comment]
I've added the type defs for the router over here: https://github.com/redwoodjs/redwood/pull/1110

**Comment by lobosan** [comment]
Once this is finished, maybe you can coin the term TAMStack (TypeScript, API, Markdown) :D

**Comment by corbt** [comment]
I'm an outsider here really excited about Redwood. First-class Typescript support is critical to me, so I wanted to get a sense on how Typescript adoption is going within the Redwood project itself. I wrote a quick script to parse the Redwood tree at each commit and count the lines of TS/JS to see the momentum. Looks like work in TS has been catching up fast, which gets me excited! Sharing the graph here in case anyone else finds it interesting.

![Redwood Project Total LOC Over Time](https://user-images.githubusercontent.com/176426/110261170-fbec7680-7f63-11eb-92c6-8497cd8ce5b8.png)

**Comment by Tobbe** [comment]
That's a really cool graph. Thanks for sharing 🙂

**Comment by jvanbaarsen** [comment]
In the docs it says:

> Redwood doesn't use the TypeScript compiler; instead, we use Babel, which strips out the types before transpiling.

I'm wondering, are there plans to use the TSCompiler at some point? Because now we lose out on a whole lot of TSC features, correct? As in, we 100% rely on the editor to let you know if a type is invalid?

**Comment by peterp** [comment]
@jvanbaarsen I don't think we'll ever compile to `tsc` since it's a bit slow, but we will eventually be fully-type compliant and checkable by `tsc`

**Comment by peterp** [comment]
But once we're full-type compliant, nothing will stop you from using tsc. It's just that we're not setting up our own tooling to use it. We rather strip out the types and transpile to JS via babel, and via experimental esbuild support.

**Comment by peterp** [comment]
I'm happy to report that we'll be full type compliant in `v.1.0.0` of RedwoodJS!

**Comment by peterp** [comment]
Here are the last few issues that we're working on to close this issue:

- [ ] Typing of Cells: Data props passed to `Success` and props required for initialization. @peterp 
- [x] Generate types for GraphQL queries.
- [ ] Remove `squiggles` from scenarios. @dac09 
- [ ] Remove `ResolveerArgs` #2437
- [ ] Route should accept all valid JSX return types.
- [ ] Directory named modules @peterp
- [x] useQuery and useMutation to have correct types @dac09 

## Developer experience

- [ ] Ctrl+Cmd click on alias imports.
- [ ] Run `rw typecheck | rw tsc` to type-check a redwood project.

**Comment by dac09** [comment]
Closing this tracking issue! Outstanding typescript issues can be tracked on the project board https://github.com/redwoodjs/redwood/projects/2! 🚀

**Comment by thedavidprice** [comment]
## Update
Redwood [v0.33.0](https://community.redwoodjs.com/t/redwood-v0-33/2176) includes major TS updates and features. Please try it out!

Feedback and discussion are happening in [this post](https://community.redwoodjs.com/t/community-help-wanted-typescript-feedback-v0-33-x/2177) on the Forums.

**Comment by riolly** [comment]
> I'm an outsider here really excited about Redwood. First-class Typescript support is critical to me, so I wanted to get a sense on how Typescript adoption is going within the Redwood project itself. I wrote a quick script to parse the Redwood tree at each commit and count the lines of TS/JS to see the momentum. Looks like work in TS has been catching up fast, which gets me excited! Sharing the graph here in case anyone else finds it interesting.
> 
> ![Redwood Project Total LOC Over Time](https://user-images.githubusercontent.com/176426/110261170-fbec7680-7f63-11eb-92c6-8497cd8ce5b8.png)

You should switch the color.

**Comment by Philzen** [comment]
I'm currently onboarding and working through the Tutorial using a TypeScript project. I happen to run into a couple of TypeScript warnings during that. Is there already an initiative to polish the Tutorial for TypeScript?

**Comment by Tobbe** [comment]
@Philzen No, nothing specific for the tutorial. But please report what you find so we can take a look. I think it would be best if you could report each individual thing as its own issue. Then we can group them if needed.

**Comment by Philzen** [comment]
> I think it would be best if you could report each individual thing as its own issue. Then we can group them if needed.

@Tobbe Thanks for the reply – will do and refer to them here.
